geode-redis/src/main/java/org/apache/geode/redis/internal/executor/string/SetExecutor.java

```diff
@@ -35,105 +37,182 @@
+
+    throwExceptionIfUnknownParameter(optionalParametersStrings);
+    throwExceptionIfIncompatableParamaterOptions(optionalParametersStrings);
+    keepTTL = optionalParametersStrings.contains("KEEPTL");
+
+    if (optionalParametersStrings.contains("PX")) {
+      String nextParameter =
+          getNextParameter("PX", optionalParametersStrings);
+
+      if (!isANumber(nextParameter)) {
+        throw new IllegalArgumentException(ERROR_SYNTAX);
+      }
+
+      expiration = parseExpirationTime("PX", nextParameter);
+      if (expiration <= 0) {
+        throw new IllegalArgumentException(ERROR_INVALID_EXPIRE_TIME);
+      }
+
+    } else if (optionalParametersStrings.contains("EX")) {
+      String nextParameter =
+          getNextParameter("EX", optionalParametersStrings);
+      if (!isANumber(nextParameter)) {
+        throw new IllegalArgumentException(ERROR_SYNTAX);
+      }
+      expiration = parseExpirationTime("EX", nextParameter);
+      if (expiration <= 0) {
+        throw new IllegalArgumentException(ERROR_INVALID_EXPIRE_TIME);
       }
     }
 
+    if (optionalParametersStrings.contains("NX")) {
+      existsOption = SetOptions.Exists.NX;
+    } else if (optionalParametersStrings.contains("XX")) {
+      existsOption = SetOptions.Exists.XX;
+    }
     return new SetOptions(existsOption, expiration, keepTTL);
   }
 
-  private long parseExpirationTime(int index, List<byte[]> commandElems)
-      throws IllegalArgumentException {
-    String expirationString;
+  private String getNextParameter(String currentParameter,
+      List<String> optionalParametersStrings) {
+    int index = optionalParametersStrings.indexOf(currentParameter);
+    if (optionalParametersStrings.size() <= index + 1) {
+      throw new IllegalArgumentException(ERROR_SYNTAX);
+    }
+    return optionalParametersStrings.get(index + 1);
+  }
+
+  private void throwExceptionIfUnknownParameter(List<String> optionalParameters) {
+    List<String> validOptionalParamaters = Arrays.asList("EX", "PX", "NX", "XX", "KEEPTL");
+
+    List<String> parametersInQuestion =
+        optionalParameters
+            .stream()
+            .filter(parameter -> (!validOptionalParamaters.contains(parameter)))
+            .collect(Collectors.toList());
+
+    parametersInQuestion.forEach(parameter -> throwErrorIfNotANumberInExpectedPosition(
+        parameter,
+        optionalParameters));
+  }
+
+  private void throwErrorIfNotANumberInExpectedPosition(
+      String parameter,
+      List<String> optionalParameters) {
+
+    if (previousOptionIsValidAndExpectsANumber(parameter, optionalParameters)) {
+      convertToLongOrThrowException(parameter);
+    } else {
+      throw new IllegalArgumentException(ERROR_SYNTAX);
+    }
+  }
 
+  private boolean isANumber(String parameter) {
     try {
-      expirationString = Coder.bytesToString(commandElems.get(index));
-    } catch (IndexOutOfBoundsException e) {
+      Long.parseLong(parameter);
```

Comment:
I don't like that we now call parseLong twice. Once in isANumber and once in parseExpirationTime. Just have parseExpirationTime handle NumberFormatException and throw IllegalArgumentException(ERROR_NOT_INTEGER)

Reply:
(see below)